lib/Haskell/Law/Eq/Def.agda

```diff
@@ -0,0 +1,94 @@
+module Haskell.Law.Eq.Def where
+
+open import Haskell.Prim
+open import Haskell.Prim.Bool
+open import Haskell.Prim.Integer
+open import Haskell.Prim.Int
+open import Haskell.Prim.Word
+open import Haskell.Prim.Double
+open import Haskell.Prim.Tuple
+open import Haskell.Prim.Maybe
+open import Haskell.Prim.Either
+
+open import Haskell.Prim.Eq
+
+open import Haskell.Law.Bool
+open import Haskell.Law.Equality
+
+record IsLawfulEq (e : Set) ⦃ iEq : Eq e ⦄ : Set₁ where
+  field
+    isEquality : ∀ (x y : e) → Reflects (x ≡ y) (x == y)
+  
+  equality : ∀ (x y : e) → (x == y) ≡ True → x ≡ y
+  equality x y h = extractTrue ⦃ h ⦄ (isEquality x y)
+
+  nequality : ∀ (x y : e) → (x == y) ≡ False → (x ≡ y → ⊥)
+  nequality x y h = extractFalse ⦃ h ⦄ (isEquality x y)
+  
+  -- contrapositive of nequality
+  equality' : ∀ (x y : e) → x ≡ y → (x == y) ≡ True
+  equality' x y h with x == y in eq
+  ... | False = magic (nequality x y eq h)
+  ... | True = refl
+  
+  -- contrapositive of equality
+  nequality' : ∀ (x y : e) → (x ≡ y → ⊥) → (x == y) ≡ False
+  nequality' x y h with x == y in eq
+  ... | True = magic (h (equality x y eq))
+  ... | False = refl
+        
+open IsLawfulEq ⦃ ... ⦄ public
+
+-- Reflexivity: x == x = True
+eqReflexivity : ⦃ iEq : Eq e ⦄ → ⦃ IsLawfulEq e ⦄
+              → ∀ (x : e) → (x == x) ≡ True
+eqReflexivity x = equality' x x refl
+
+-- Symmetry: x == y = y == x
+eqSymmetry : ⦃ iEq : Eq e ⦄ → ⦃ IsLawfulEq e ⦄
+           → ∀ (x y : e) → (x == y) ≡ (y == x)
+eqSymmetry x y with x == y in eq
+... | True  = sym (equality' y x (sym (equality x y eq)))
+... | False = sym (nequality' y x (λ qe → (nequality x y eq) (sym qe)))
+
+-- Transitivity: if x == y && y == z = True, then x == z = True
+eqTransitivity : ⦃ iEq : Eq e ⦄ → ⦃ IsLawfulEq e ⦄
+               → ∀ (x y z : e) → ((x == y) && (y == z)) ≡ True → (x == z) ≡ True
+eqTransitivity x y z h 
+  = equality' x z (trans 
+    (equality x y (&&-leftTrue (x == y) (y == z) h))
+    (equality y z (&&-rightTrue (x == y) (y == z) h)))
+
+-- Extensionality: if x == y = True and f is a function whose return type is an instance of Eq, then f x == f y = True
```

Comment:
This is not extensionality but congruence.

Reply:
It seems like congruence is specifically for the function application, but I think here with the equality, extensionality actually makes sense for this property: "[The axiom of extensionality] says that sets having the same elements are the same set."

(I'm not very attached to this name, so I can still change it, but I mostly think it's nice to keep with the naming scheme on [Hackage](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Eq.html).)